File: app/src/main/java/com/chiranths/jobportal1/Activities/OtpLoginActivity.java
package com.chiranths.jobportal1.Activities;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.chiranths.jobportal1.Activities.BasicActivitys.UserDetailsActivity;
import com.chiranths.jobportal1.Activities.Dashboard.StartingActivity;
import com.chiranths.jobportal1.Activities.jobs.MainActivity;
import com.chiranths.jobportal1.R;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.FirebaseException;
import com.google.firebase.FirebaseTooManyRequestsException;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseAuthInvalidCredentialsException;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.PhoneAuthCredential;
import com.google.firebase.auth.PhoneAuthOptions;
import com.google.firebase.auth.PhoneAuthProvider;

import java.util.concurrent.TimeUnit;

public class OtpLoginActivity extends AppCompatActivity implements View.OnClickListener {

    EditText edtPhone,edt_otp;
    TextView tvSendOtp;
    CardView cvOtp;
    Button btn_continue;

    FirebaseAuth mAuth;
    String mVerificationId;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_otp_login);

        mAuth = FirebaseAuth.getInstance();

        edtPhone= findViewById(R.id.edt_phone);
        tvSendOtp= findViewById(R.id.tv_send_otp);
        cvOtp= findViewById(R.id.cv_otp);
        btn_continue= findViewById(R.id.btn_continue);
        edt_otp= findViewById(R.id.edt_otp);
        tvSendOtp.setOnClickListener(this);
        btn_continue.setOnClickListener(this);

        edtPhone.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (s.length() == 10) {
                    tvSendOtp.setVisibility(View.VISIBLE);
                } else {
                    tvSendOtp.setVisibility(View.GONE);
                    cvOtp.setVisibility(View.GONE);
                }

            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

    }

    @Override
    public void onClick(View view) {

        switch (view.getId())
        {

            case R.id.tv_send_otp:
                cvOtp.setVisibility(View.VISIBLE);

                // OnVerificationStateChangedCallbacks

                PhoneAuthOptions options =
                        PhoneAuthOptions.newBuilder(mAuth)
                                .setPhoneNumber("+91" + edtPhone.getText().toString())            // Phone number to verify
                                .setTimeout(60L, TimeUnit.SECONDS) // Timeout and unit
                                .setActivity(this)                 // Activity (for callback binding)
                                .setCallbacks(mCallbacks)           // OnVerificationStateChangedCallbacks
                                .build();
                PhoneAuthProvider.verifyPhoneNumber(options);
                break;

            case R.id.btn_continue:

                if (edt_otp.length() == 0) {
                    Toast.makeText(getApplicationContext(),"Enter the otp", Toast.LENGTH_SHORT).show();
                }
                else
                {
                    PhoneAuthCredential credential = PhoneAuthProvider.getCredential(mVerificationId, edt_otp.getText().toString());
                    signInWithPhoneAuthCredential(credential);
                }

                break;
        }
    }

    PhoneAuthProvider.OnVerificationStateChangedCallbacks mCallbacks = new PhoneAuthProvider.OnVerificationStateChangedCallbacks() {

        @Override
        public void onVerificationCompleted(PhoneAuthCredential credential) {
            // This callback will be invoked in two situations:
            // 1 - Instant verification. In some cases the phone number can be instantly
            //     verified without needing to send or enter a verification code.
            // 2 - Auto-retrieval. On some devices Google Play services can automatically
            //     detect the incoming verification SMS and perform verification without
            //     user action.
            Log.d("TAG", "onVerificationCompleted:" + credential);

            signInWithPhoneAuthCredential(credential);
        }

        @Override
        public void onVerificationFailed(FirebaseException e) {


            Toast.makeText(getApplicationContext(),"failed", Toast.LENGTH_SHORT).show();


            Log.w("TAG", "onVerificationFailed", e);

            if (e instanceof FirebaseAuthInvalidCredentialsException) {

            } else if (e instanceof FirebaseTooManyRequestsException) {

            }

        }

        @Override
        public void onCodeSent(@NonNull String verificationId,
                               @NonNull PhoneAuthProvider.ForceResendingToken token) {
            // The SMS verification code has been sent to the provided phone number, we
            // now need to ask the user to enter the code and then construct a credential
            // by combining the code with a verification ID.
            Log.d("TAG", "onCodeSent:" + verificationId);

            // Save verification ID and resending token so we can use them later
            mVerificationId = verificationId;

            // ...
        }


    };

    private void signInWithPhoneAuthCredential(PhoneAuthCredential credential) {
        mAuth.signInWithCredential(credential)
                .addOnCompleteListener(this,new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (task.isSuccessful()) {

                            startActivity(new Intent(OtpLoginActivity.this, StartingActivity.class));
                            finish();

                            // FirebaseUser user = task.getResult().getUser();

                        } else {

                            Toast.makeText(getApplicationContext(), "Invalid Otp", Toast.LENGTH_SHORT).show();

                           /* Log.w("TAG", "signInWithCredential:failure", task.getException());
                            if (task.getException() instanceof FirebaseAuthInvalidCredentialsException) {
                                // The verification code entered was invalid
                            }*/
                        }
                    }
                });
    }
}

File: app/src/main/java/com/chiranths/jobportal1/Activities/OtpLoginActivity.kt
package com.chiranths.jobportal1.Activities

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.cardview.widget.CardView
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.chiranths.jobportal1.Activities.BasicActivitys.UserDetailsActivity
import com.chiranths.jobportal1.R
import com.chiranths.jobportal1.Utilitys.AppConstants
import com.chiranths.jobportal1.Utilitys.PreferenceManager
import com.google.firebase.FirebaseException
import com.google.firebase.FirebaseTooManyRequestsException
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseAuthInvalidCredentialsException
import com.google.firebase.auth.PhoneAuthCredential
import com.google.firebase.auth.PhoneAuthProvider
import com.google.firebase.auth.PhoneAuthProvider.ForceResendingToken
import com.google.firebase.auth.PhoneAuthProvider.OnVerificationStateChangedCallbacks
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import java.util.concurrent.TimeUnit

class OtpLoginActivity : AppCompatActivity(), View.OnClickListener {
    var edtPhone: EditText? = null
    var edt_otp: EditText? = null
    var tvSendOtp: TextView? = null
    var cvOtp: CardView? = null
    var btn_continue: Button? = null
    var mVerificationId: String? = null
    lateinit var preferenceManager: PreferenceManager
    private lateinit var mAuth: FirebaseAuth
    private var verificationId: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_otp_login)
        preferenceManager= PreferenceManager(this);
        mAuth = FirebaseAuth.getInstance()
        edtPhone = findViewById(R.id.edt_phone)
        tvSendOtp = findViewById(R.id.tv_send_otp)
        cvOtp = findViewById(R.id.cv_otp)
        btn_continue = findViewById(R.id.btn_continue)
        edt_otp = findViewById(R.id.edt_otp)
        tvSendOtp?.setOnClickListener(this)
        btn_continue?.setOnClickListener(this)
        edtPhone?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (s.length == 10) {
                    tvSendOtp?.setVisibility(View.VISIBLE)
                } else {
                    tvSendOtp?.setVisibility(View.GONE)
                    cvOtp?.setVisibility(View.GONE)
                }
            }

            override fun afterTextChanged(s: Editable) {}
        })
    }








    override fun onClick(view: View) {
        when (view.id) {
            R.id.tv_send_otp -> {
                cvOtp?.visibility = View.VISIBLE
                val phoneNumber = "+91" + edtPhone?.text.toString().trim { it <= ' ' }
                // OnVerificationStateChangedCallbacks
                PhoneAuthProvider.getInstance().verifyPhoneNumber(
                    phoneNumber,  // Phone number to verify
                    60,  // Timeout duration
                    TimeUnit.SECONDS,  // Unit of timeout
                    this,  // Activity (for callback binding)
                    mCallbacks
                )
            }
            R.id.btn_continue ->
                if (edt_otp?.text?.length == 0) {

                Toast.makeText(applicationContext, "Enter the otp", Toast.LENGTH_SHORT).show()
                /*val sharedPreferences = getSharedPreferences("MySharedPref", MODE_PRIVATE)
                val myEdit = sharedPreferences.edit()
                preferenceManager.saveLoginState(true)
                myEdit.putString(AppConstants.number, edtPhone?.text.toString())
                myEdit.commit()
                //fetchProfile(edtPhone?.text.toString())
                startActivity(Intent(this@OtpLoginActivity, UserDetailsActivity::class.java))*/

            } else if (edt_otp?.text?.length == 6) {
                verifyCode(edt_otp?.text.toString())
            }
        }
    }

    var mCallbacks: OnVerificationStateChangedCallbacks =
        object : OnVerificationStateChangedCallbacks() {
            override fun onVerificationCompleted(credential: PhoneAuthCredential) {
                val code = credential.smsCode
                if (code != null) {
                    edt_otp!!.setText(code)
                }
            }

            override fun onVerificationFailed(e: FirebaseException) {
                Toast.makeText(applicationContext, "failed", Toast.LENGTH_SHORT).show()
                Log.w("TAG", "onVerificationFailed", e)
                if (e is FirebaseAuthInvalidCredentialsException) {
                } else if (e is FirebaseTooManyRequestsException) {
                }
            }

            override fun onCodeSent(
                verificationId: String,
                token: ForceResendingToken
            ) {
                Log.d("TAG", "onCodeSent:$verificationId")
                mVerificationId = verificationId
            }
        }

    // below method is use to verify code from Firebase.
    private fun verifyCode(code: String) {
        val credential = PhoneAuthProvider.getCredential(mVerificationId!!, code)
        signInWithPhoneAuthCredential(credential)
    }

    private fun signInWithPhoneAuthCredential(credential: PhoneAuthCredential) {
        mAuth!!.signInWithCredential(credential)
            .addOnCompleteListener(this) { task ->
                if (task.isSuccessful) {
                    /*val sharedPreferences = getSharedPreferences("MySharedPref", MODE_PRIVATE)
                    val myEdit = sharedPreferences.edit()
                    myEdit.putString(AppConstants.number, edtPhone?.text.toString())
                    myEdit.commit()*/
                    fetchProfile(edtPhone?.text.toString())

                } else {
                    Toast.makeText(applicationContext, "Invalid Otp", Toast.LENGTH_SHORT).show()
                }
            }
    }

    private fun fetchProfile(userNumber: String) {
        if (userNumber !== "") {
            val profile = FirebaseDatabase.getInstance().reference.child(AppConstants.profile).child(userNumber)
            profile.addValueEventListener(object : ValueEventListener {
                override fun onDataChange(snapshot: DataSnapshot) {
                    if (snapshot.exists()) {
                        val dataMap = snapshot.value as HashMap<String, Any>?
                        for (key in dataMap!!.keys) {
                            val sharedPreferences = getSharedPreferences("MySharedPref", MODE_PRIVATE)
                            val myEdit = sharedPreferences.edit()
                            myEdit.putString(AppConstants.number, edtPhone?.text.toString())
                            myEdit.commit()
                            finish()
                        }
                    } else {
                        startActivity(Intent(this@OtpLoginActivity, UserDetailsActivity::class.java))
                        finish()
                    }
                }
                override fun onCancelled(error: DatabaseError) {}
            })
        }
    }

}